librad/src/id.rs

```diff
@@ -0,0 +1,273 @@
+use hex::{decode, encode};
+use multihash::{Multihash, Sha2_256};
+use std::collections::{HashMap, HashSet};
+//use serde::{Deserialize, Serialize};
+//use serde_json;
+use crate::{
+    keys::device::{Key, PublicKey, Signature},
+    peer::PeerId,
+};
+
+pub mod data;
+
+#[derive(Debug, Fail)]
+pub enum Error {
+    #[fail(display = "Serialization failed")]
+    SerializationFailed(serde_json::error::Error),
+
+    #[fail(display = "Invalid UTF8")]
+    InvalidUtf8(std::string::FromUtf8Error),
+
+    #[fail(display = "Invalid hex")]
+    InvalidHex(String),
+
+    #[fail(display = "Invalid hash")]
+    InvalidHash(String),
+
+    #[fail(display = "Signature already present")]
+    SignatureAlreadyPresent(PublicKey),
+
+    #[fail(display = "Key not present")]
+    KeyNotPresent(PublicKey),
+
+    #[fail(display = "User not present")]
+    UserNotPresent(RadicleUri),
+
+    #[fail(display = "User key not present")]
+    UserKeyNotPresent(RadicleUri, PublicKey),
+
+    #[fail(display = "Signature missing")]
+    SignatureMissing,
+
+    #[fail(display = "Signature decoding failed")]
+    SignatureDecodingFailed,
+
+    #[fail(display = "Signature verification failed")]
+    SignatureVerificationFailed,
+
+    #[fail(display = "Resolution failed")]
+    ResolutionFailed(String),
+}
+
+#[derive(Clone, Debug)]
+pub struct RadicleUri {
```

Comment:
This is the relevant bit for #65 -- what I dubbed there as "physical" address is the commented part. I'm wondering if the "logical" part requires a tag conveying what kind of data to expect when resolving it. 

On paper, it could pan out nicely if what is called a `RadicleUri` here is a URN, e.g.:

`urn:rad:ident:user:<multihash>`

which may be resolved to one or more URLs, e.g.

`rad://PEER_ID/PROJECT/branch#MULTIHASH`

(not sure about the file name)